# Lecture 4.3: Case Study

<div style="border: 1px double black; padding: 10px; margin: 10px">

**Goals for this lecture:**
* Use the tools we have learnt so far to study more about baseball data set
    
</div>



## Who is the greatest batter of all time?
The `Lahman` dataset contains information on baseball players.

In [3]:
install.packages("Lahman")


The downloaded binary packages are in
	/var/folders/br/gjgfqmdn1d5cmwrq6b8z_mf80000gn/T//RtmpWf91sq/downloaded_packages


In [1]:
#options(repr.plot.width=6, repr.plot.height=4)
library(tidyverse)


── Attaching packages ─────────────────────────────────────── tidyverse 1.3.0 ──

✔ ggplot2 3.3.2     ✔ purrr   0.3.4
✔ tibble  3.0.3     ✔ dplyr   1.0.2
✔ tidyr   1.1.1     ✔ stringr 1.4.0
✔ readr   1.3.1     ✔ forcats 0.5.0

── Conflicts ────────────────────────────────────────── tidyverse_conflicts() ──
✖ dplyr::filter() masks stats::filter()
✖ dplyr::lag()    masks stats::lag()



In [4]:
library(Lahman)
bat <- as_tibble(Batting) %>% print

# A tibble: 107,429 x 22
   playerID yearID stint teamID lgID      G    AB     R     H   X2B   X3B    HR
   <chr>     <int> <int> <fct>  <fct> <int> <int> <int> <int> <int> <int> <int>
 1 abercda…   1871     1 TRO    NA        1     4     0     0     0     0     0
 2 addybo01   1871     1 RC1    NA       25   118    30    32     6     0     0
 3 allisar…   1871     1 CL1    NA       29   137    28    40     4     5     0
 4 allisdo…   1871     1 WS3    NA       27   133    28    44    10     2     2
 5 ansonca…   1871     1 RC1    NA       25   120    29    39    11     3     0
 6 armstbo…   1871     1 FW1    NA       12    49     9    11     2     1     0
 7 barkeal…   1871     1 RC1    NA        1     4     0     1     0     0     0
 8 barnero…   1871     1 BS1    NA       31   157    66    63    10     9     0
 9 barrebi…   1871     1 FW1    NA        1     5     1     1     1     0     0
10 barrofr…   1871     1 BS1    NA       18    86    13    13     2     1     0
# … with 107,41

Each row of the above data is data per player and per year.  For instance, let's take a look at the second row of our data.

Let's try to compute the career batting average for Bob Addy, defined as Hit/(Number of bats).  The playerID for Bob Addy is addybo01. 

Bob Addy was active in the years 1871-1877. During that time he had $118+51+152+213+310+142+245=1231$ at-bats, and $32+16+54+51+80+40+68=341$ hits. Therefore his career batting average was $341/1241=0.277$.

### Exercise
By appropriately grouping and summarizing the data, add up all the hits and at-bats for each player across all the years they played, and compute their career batting average. 

Which player(s) has the highest career batting average?

### Always include counts
It is a good idea to include counts of each group when you do a summary. Some groups may have very low numbers of observations, resulting in high variance for the summary statistics. 



What happens if we restrict our batting average calculation to players that had at least 100 at-bats, and sort it from players that have the highest batting average?

What is wrong with the above?  Why are there so many overlapping rows for the player cobbty01? 

### Exercise
Output all the years in which the batting average for those years for cobbty01 is less than his bat_avg across all years. That is, we are interested in knowing which year cobbty01 underperform.     

## Exercise
Now output all rows for each player that underperform, that is, their bat_avg_year is less than bat_avg.

That is, we are interested in knowing whether there is any specific year that the players underperform. 

## Exercise
Write code to find the player with the highest bat_avg for each team. You may find the function `slicemax` useful. 

## Names of baseball players

Let us think more about names. Naming frequencies change a lot over time. There are 19617 baseball players in this data set. How have their names changed over time?

We are going to extract the first name and last name in our data set `Lahman::Master` by linking the playerID.

## Exercise
What were the top five most common first names for players born before 1900? After 1980?

One thing we notice is that there are a lot of nicknames. It might make more sense to look at the "given name", which is usually the first and middle names. To do this, we will need to split up these names. There is a built-in command for doing this in R:

## Exercise
What were the top five most common first names for players born before 1900? After 1980?

## Finding distinct values

Here's an example: I want to know how many distinct values are there? The `n_distinct()` function takes a vector of values, and returns the number of distinct values:

In [85]:
n_distinct(c("Michigan", "Ohio", "STATS 306", "Michigan"))

[1] 3

## Exercise
How many distinct names were there among players born before 1900? After 1980?

Are there more or less unique names now than there were in the past? Let's consider the number of distinct names seen in each year: